lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp

```diff
@@ -34,6 +34,49 @@ Value getWaveN(ConversionPatternRewriter &rewriter, Location loc, Value wave,
 
 namespace SharedToDotOperandMFMA {
 
+/**
+ * @brief swizzling tensor element indexes according pattern encoded in
+ * SharedEncodingAttr
+ *
+ * @param rewriter
+ * @param loc
+ * @param row row of target tensor element related to the start of smemObj
+ * @param col col of target tensor element related to the start of smemObj
+ * @param smemObj shared memory object, contains info about tensor in LDS
+ * @param attr layout attribute, contains swizzling info
+ * @return swizzled row, col indexes in tensor notation
+ */
+std::pair<mlir::Value, mlir::Value>
+swizzleIndexes(ConversionPatternRewriter &rewriter, Location loc, Value row,
+               Value col, SharedMemoryObject smemObj, SharedEncodingAttr attr) {
+  (void)smemObj; // unused in current pattern
+  bool transposed = (attr.getOrder()[0] != 1);
+  if (transposed) {
+    // tensor is column-wise, so swapping col and row in computations
+    std::swap(row, col);
+  }
+  auto vec = i32_val(attr.getVec());
+  auto perPhase = i32_val(attr.getPerPhase());
+  auto maxPhase = i32_val(attr.getMaxPhase());
+
+  // Original algorithm taken from getSwizzledSharedPtrs function
+  // (TritonGPUToLLVMBase.h): Basic algorithm for row-major tensor is following:
+  //
+  // phase = (row // perPhase) % maxPhase
+  // colOffSwizzled = ((col // vec) ^ phase) * vec
+  // colOffOrdered = col % vec
```

Comment:
IIRC, the upstream way compute `colOffOrdered` based on `minVec`.
So why there is only `vec` here?

Reply:
minVec is used for loading/storing **vectors** of elements in one operation.
Current implementation of mfma dot load every element separately.

you can see, that actual swizzling uses only vec in original code as well, inVec value is used only for alignment: https://github.com/ROCmSoftwarePlatform/triton/blob/triton-mlir/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h#L344

Reply:
I think `minVec` is used for storing vectors to lds in one instruction. When it comes loading data from LDS for mfma, it should use `vec`.
OK I think we need to talk about `vec`, `minVec`, `inVec`, and `outVec` since they are used when moving data to/from LDS.
- `inVec` is the vector length each thread can load during one **global_load** instruction. It can be inferred from the BlockedEncodingAttr of the tensor, which is done in [storeDistributedToShared()](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp#L560). And [here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h#L431) shows how `inVec` is set. Note that if the leading dims from the src and dst does not match, `inVec` is set to 1, meaning each thread can only load 1 element at a time.
- `outVec` is the same as `vec`, which is obtained from the sharedEncodingAttr. This is the vector length each thread can load during one **lds_load** (`ds_read_xxx` instruction). This is determined by the mfma input layout.
- `inVec` and `outVec` can be different and `minVec` is set to `min(inVec, outVec)`, which is used as the vector length each thread can write during one **lds_store** (`ds_write` instructions). Here is the [loop](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h#L452) for LDS write.

When computing swizzling, we want to swizzle elements in LDS as a group of `vec`/`outVec` elements, since threads in the warps can access `vec` columns of elements at the same time. This is why `colOffSwizzled` is based on `vec`.
Because `minVec` elements are stored at a time, we want to compute the swizzled address for every `minVec` elements. This is why `colOffOrdered` is based on `minVec` as [here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h#L266).

Reply:
Everything is right, except the last statement:
> Because minVec elements are stored at a time, we want to compute the swizzled address for every minVec elements.

This is true for `distributedToShared` and, probably, `sharedToMMA` conversions.
Currently `sharedToMFMA` do not load values in chunks.

I would like to return to this later, after we reduce bank conflicts, do proper profiling of generated code and find if  loading several values in one instruction gives us an advantage.

Reply:
> Currently sharedToMFMA do not load values in chunks

So this is the reason everyone (@weihanmines  and Raman) only sees `ds_read_b16` in the generated code.
rocMLIR and CK always load enough number of elements from LDS for mfma instructions. And Triton's `ldmartix` instruction does the same. I don't see why it can be beneficial to load one element per instruction.

Reply:
> Current implementation of mfma dot load every element separately.

I still want to clarify about this part. It seems to me one-element-at-a-time is true when loading data from LDS to VGPR for mfma instructions. Then what about storing data to LDS? Does each thread still store one element at a time?
- If so, `vec`/`minVec` is still assumed to be 1 in this PR
- If not, then each thread stores `minVec` elements at a time. Then the offset computation needs to be done for every `minVec` elements instead of `vec`